### PR TITLE
add support for mips(el)-unknown-linux-musl

### PR DESCRIPTION
**Pull request by japaric**
This target covers MIPS devices that run the trunk version of OpenWRT.

The x86_64-unknown-linux-musl target always links statically to C libraries. For
the mips(el)-unknown-linux-musl target, we opt for dynamic linking (like most of
other targets do) to keep binary size down.

As for the C compiler flags used in the build system, we use the same flags used
for the mips(el)-unknown-linux-gnu target.

r? @alexcrichton

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -16,4 +16,5 @@
 	url = https://github.com/rust-lang/rust-installer.git
 [submodule "src/liblibc"]
 	path = src/liblibc
-	url = https://github.com/rust-lang-nursery/libc.git
+	url = https://github.com/japaric/libc.git
```
Comment (japaric) [inline .gitmodules]: TODO revert this change

File: mk/cfg/mips-unknown-linux-musl.mk [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# mips-unknown-linux-musl configuration
+CC_mips-unknown-linux-musl=mips-openwrt-linux-gcc
```
Comment (japaric) [inline mk/cfg/mips-unknown-linux-musl.mk]: TODO change the prefixes to something vendor neutral. Probably `mips-linux-musl-gcc`.

File: src/liblibc [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit 91ff43c736de664f8d3cd351e148c09cdea6731e
+Subproject commit 81c05bcddb05d0e5facf66f3945405bbf50b04d3
```
Comment (japaric) [inline src/liblibc]: TODO this will need to be updated

File: src/libstd/os/linux/raw.rs [hunk 4]
```diff
@@ -93,11 +93,20 @@ mod arch {
     use os::raw::{c_long, c_ulong};
     use os::unix::raw::{gid_t, uid_t};
 
+    #[cfg(target_env = "musl")]
+    #[stable(feature = "raw_ext", since = "1.1.0")] pub type blkcnt_t = i64;
```
Comment (japaric) [inline src/libstd/os/linux/raw.rs]: I'm not sure what stability level this should have. AFAIK, these types are going to be deprecated in the future.

File: src/librustc_back/target/mipsel_unknown_linux_musl.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// except according to those terms.
+
+use target::Target;
+
+pub fn target() -> Target {
+    Target {
+        llvm_target: "mipsel-unknown-linux-musl".to_string(),
+        target_endian: "little".to_string(),
+        target_pointer_width: "32".to_string(),
+        arch: "mips".to_string(),
```
Comment (alexcrichton) [inline src/librustc_back/target/mipsel_unknown_linux_musl.rs]: Should this be mipsel?

File: mk/cfg/mipsel-unknown-linux-musl.mk [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# mipsel-unknown-linux-musl configuration
+CC_mipsel-unknown-linux-musl=mipsel-openwrt-linux-gcc
+CXX_mipsel-unknown-linux-musl=mipsel-openwrt-linux-g++
+CPP_mipsel-unknown-linux-musl=mipsel-openwrt-linux-gcc
+AR_mipsel-unknown-linux-musl=mipsel-openwrt-linux-ar
```
Comment (alexcrichton) [inline mk/cfg/mipsel-unknown-linux-musl.mk]: This should also work with a standard `mipsel-linux-gnu-gcc` toolchain, right?

Comment (alexcrichton) [inline mk/cfg/mipsel-unknown-linux-musl.mk]: Oh, I see your TODO above

File: src/libstd/os/linux/raw.rs (hunk 4 above)
Comment (alexcrichton) [inline src/libstd/os/linux/raw.rs]: Ah yeah these are fine to stay as-is

**Comment by alexcrichton** [comment]
Looks good to me, thanks @japaric!

This brings up an interesting facet about the MUSL targets, however, about whether the C standard library is statically or dynamically linked. The `x86_64-unknown-linux-musl` target was originally created with the intention of only really being used for statically linked binaries. I'd personally prefer that these targets remain consistent, but it seems that the current primary use case (OpenWRT?) is dynamically linking the C standard library.

In that sense all of the MUSL targets should really support *either* dynamically linking or statically linking, but we don't really support this sort of lazy linkage in the standard library (only on crates.io where you build from source).

I think for now I'd be fine basically choosing "whatever seems best" for the target at hand (e.g. static for x86_64 and perhaps i686 but dynamic for mips), but it's something we may wish to reconsider in the future.

**Comment by alexcrichton** [comment]
ah and of course cc @rust-lang/tools about the dynamic-vs-static linkage of musl

File: src/librustc_back/target/mipsel_unknown_linux_musl.rs (hunk 5 above)
Comment (japaric) [inline src/librustc_back/target/mipsel_unknown_linux_musl.rs]: > Should this be mipsel?

Not sure! The existing `mipsel-unknown-linux-gnu` target is also using `mips` as the arch.

AIUI, this arch field is used for `#[cfg(target_arch = ...)]` and not for codegen (that's what llvm-target is used for). If we want to change both targets (mipsel-gnu and mipsel-musl) to use arch = mipsel, then we'll have to update the cfgs in crates like libc from `#[cfg(target_arch = "mips")]` to `#[cfg(any(target_arch = "mips", any(target_arch = "mipsel"))]`.

Comment (alexcrichton) [inline src/librustc_back/target/mipsel_unknown_linux_musl.rs]: Oh interesting! That may be a bug in the target triple, however, depending on how you look at it. Either that or we may just want to make sure that we're consistent. Some data points:

* The newly-added `powerpc64le-unknown-linux-gnu` triple has `powerpc64le` as the `target_arch`.
* There's some `cfg(target_arch = "mipsel")` directives in the standard library, which apparently are doing nothing.
* We don't necessarily require the arch in the triple to equal the arch in `target_arch`, for example `armv7` recently added has the `target_arch` of `arm`.

I think that I'd prefer to lean towards leaving this as `mips` and tweaking the `powerpc64le` triple to have `powerpc64` as the `target_arch`. That seems the most consistent to me and also the most ergonomic. The two triples can still be differentiated by looking at `target_endian`

Comment (japaric) [inline src/librustc_back/target/mipsel_unknown_linux_musl.rs]: > I think that I'd prefer to lean towards leaving this as mips and tweaking the powerpc64le triple to have powerpc64 as the target_arch. That seems the most consistent to me and also the most ergonomic. The two triples can still be differentiated by looking at target_endian

:+1:  from me

**Comment by japaric** [comment]
This is now merge ready.

**Comment by japaric** [comment]
The failed test (os::raw::tests::unix) was caused by rust-lang-nursery/libc#122 which changed the definition of nlink_t to usize. I'll send a PR to libc to restore the old definition.
